### DISCUSSION TITLE
Joining 2 commands into a single one liner help please

### DISCUSSION THREAD

**Discussion by jaffamuffin** [ImageMagick/ImageMagick#5655] [Help] [answered]
Hello

using imagemagick ImageMagick 7.0.10-22 Q8 x64 2020-06-27  on win10

I want to compose one image ontop of another image and then resize to a given size. : 

`magick background.jpg top.jpg -resize 2330x3307 -gravity center -composite -quality 100 out.jpg `

I don't want the hard edge that happens so I have another commmand that can blur the image edges.

`magick top.jpg -bordercolor black -fill white ^( -clone 0 -colorize 100 -shave 10x10 -border 10x10 -blur 0x10 ^) -compose copyopacity -composite top-edgeblur.png`

Then I could use this new blurred image to compose onto my background. 

`magick background.jpg top-edgeblur.png -resize 2330x3307 -gravity center -composite -quality 100 out.jpg `

This works. 

But is there a way to have this run as one command ? 

`magick background.jpg ^( top.jpg -bordercolor black -fill -white ^( -clone 0 -colorize 100 -shave 10x10 -border 10x10 -blur 0x10 ^) -compose copyopacity -composite ^) -resize 2330x3307 -gravity center -composite -compose multiply -quality 100 out.jpg`

I tried this but it does not work.  magick: unable to open image '-white':

Any help appreciated thanks.

**Comment by GeeMack** [comment]
The argument to your `-fill` setting is "-white" beginning with a dash. You should just use "white".

**Comment by snibgo** [comment]
As GeeMack says.

Also, the end of the "one command" is:
```
... -composite -compose multiply -quality 100 out.jpg
```

But setting "-compose" *after* composite won't do anything. The default "-compose" setting is "over", but that has been changed by "-compose copyopacity". So I suggest changing the ending to:
```
... -compose over -composite -quality 100 out.jpg
```

**Reply by jaffamuffin** [reply]
Thank you - working well now. I bad typo late at night otherwise I was pretty close.

Can I check that in the magick command line, just putting a 'subcommand' in parentheses (like above) is really like unix backticks -- magick will evaluate between the parentheses and then substitute that as an image into the parent command ?    

Also seem to be able only to select one reply as 'answer' but both of you have helped thanks.

**Comment by snibgo** [comment, marked as answer]
> ... just putting a 'subcommand' in parentheses (like above) is really like unix backticks ...?

No, it is different. Well, kind of similar, but somewhat different.

Unix backtick subcommands are executed first, with the output substituted in the outer command.

IM v7 options are executed strictly in the order given. "(" and ")" are options.

We can break a command into lines for clarity. Windows syntax:

```
magick ^
background.jpg ^
^( ^
   top.jpg -bordercolor black -fill -white ^
   ^( ^
      -clone 0 -colorize 100 -shave 10x10 -border 10x10 -blur 0x10 ^
   ^) ^
   -compose copyopacity -composite ^
^) ^
-resize 2330x3307 -gravity center -compose Over -composite ^
-quality 100 out.jpg
```
This reads background.jpg, so the current image list has one entry.

"(" puts the current image list on a stack, and starts a new empty list.

We read top.jpg, so the current list again has one image. Then we set a couple of settings.

"(" again puts the current image list on a stack, and starts a new empty list.

"-clone 0" copies the first image from the list at the top of the stack. That list contains only the image from top.jpg, so that is copied into the current list.

Then we colorize all the images in the current list. (There is only one image.) And we shave it, border it, and blur it. The colour used for the border has already been set to black.

")" closes the current list, copies any images into it to the list at the top of stack, and pops the list from the stack. So now the list contains two images: top.jpg, and a processed version of top.jpg.

"-composite" replaces the two images with a single image.

The final ")" again closes the current list, copies any images into it to the list at the top of stack, and pops the list from the stack. So now the list contains two images: background.jpg, and a processed version of top.jpg.

All the images in the current list (there are two) are resized. "-composite" replaces them with one image.

That image is written to out.jpg.

We can see that the operations must be executed in the order given. Otherwise, if "( -clone 0 ...)" was executed first, there would be nowhere for the clone to be copied from.

Writing it all out makes it seem complicated. The main concepts are: we have a current list, and a stack of image lists. "(" pushes the current list to the stack, and ")" pops a list from the stack. "(" starts a new empty list. ")" copies any images that were on the list to the end of the popped list.

**Reply by jaffamuffin** [reply]
Thanks that's a very helpful reply I think it clearly shows how it works.  I wasn't quite aware of this 'stack' idea but it seems to make sense.
